api/numaresourcesoperator/v1/helper/nodegroup/nodegroup.go

```diff
@@ -29,17 +29,14 @@ import (
 )
 
 type Tree struct {
-	NodeGroup          *nropv1.NodeGroup
-	MachineConfigPools []*mcov1.MachineConfigPool
+	NodeGroup         *nropv1.NodeGroup
+	MachineConfigPool *mcov1.MachineConfigPool
```

Comment:
I'd like to merge this change so much, but we need to make sure len(MachineConfigPools) == 1 always holds true. Are we fully confident this is the case?

Reply:
Offline I wrote (and I'm reiterating here) I can't imagine a usecase on which len(MachineConfigPools) > 1 and I can't imagine a valid usecase for this. But still, this is what we had till now so if we want to make stricter (which I'd VERY MUCH like to do!) we need to take extra care to not break possible users.
If we manage to demonstrate len(MachineConfigPools) == 1 alwyas carefully analyzing the code and the possible flows, I'm all to merge this change.
I think this should be done in a separate (and welcome) PR though

Reply:
I want this change very badly, but we need to tread carefully. So what I'd do is

1. Add a `NodeGroupName` back in the `MachineConfigPool` struct so we can now backlink MachineConfigPools to their NodeGroup should a NodeGroup match more than a MachineConfigPool. It really should not, if does it's unintended and a bug, but still
2.  *obviously in another PR* keep the NodeGroupTree as-is, but if we discover a NodeGroup matches more than a MachineConfigPool, do everything like today, but set the status.Conditions as Degraded, reporting that `NodeGroup XYZ matches N MachineConfigPool, which is unsupported. Please fix the config.
3. Optionally (but I'm inclined to do it) add a magic annotation to silence the warning.
4. After 2 releases (4.20), actually apply this change right here and force 1:1 between NodeGroups and MachineConfigPools

this is clumsy and unfortunate, but in order to maximize the backward compatibility and fix our code while keeping users happy, it seems the best we can do atm.